components/web_discovery/browser/double_fetcher.cc

```diff
@@ -0,0 +1,132 @@
+      data: "Page data"
+      destination: WEBSITE
+    }
+    policy {
+      cookies_allowed: NO
+      setting:
+        "Users can opt-in or out via brave://settings/search"
+    })");
+
+}  // namespace
+
+DoubleFetcher::DoubleFetcher(
+    PrefService* profile_prefs,
+    network::SharedURLLoaderFactory* shared_url_loader_factory,
+    FetchedCallback callback)
+    : profile_prefs_(profile_prefs),
+      shared_url_loader_factory_(shared_url_loader_factory),
+      request_queue_(profile_prefs,
+                     kScheduledDoubleFetches,
+                     kRequestMaxAge,
+                     kMinRequestInterval,
+                     kMaxRequestInterval,
+                     kMaxRetries,
+                     base::BindRepeating(&DoubleFetcher::OnFetchTimer,
+                                         base::Unretained(this))),
+      callback_(callback) {}
+
+DoubleFetcher::~DoubleFetcher() = default;
+
+void DoubleFetcher::ScheduleDoubleFetch(const GURL& url,
+                                        base::Value associated_data) {
+  base::Value::Dict fetch_dict;
+  fetch_dict.Set(kUrlKey, url.spec());
+  fetch_dict.Set(kAssociatedDataKey, std::move(associated_data));
+
+  request_queue_.ScheduleRequest(std::move(fetch_dict));
+}
+
+void DoubleFetcher::OnFetchTimer(const base::Value& request_data) {
+  const auto* fetch_dict = request_data.GetIfDict();
+  const auto* url_str = fetch_dict ? fetch_dict->FindString(kUrlKey) : nullptr;
+  if (!url_str) {
+    request_queue_.NotifyRequestComplete(true);
+    return;
+  }
+
+  GURL url(*url_str);
+  auto resource_request = CreateResourceRequest(url);
+  url_loader_ = network::SimpleURLLoader::Create(
+      std::move(resource_request), kFetchNetworkTrafficAnnotation);
+  url_loader_->DownloadToString(
+      shared_url_loader_factory_.get(),
+      base::BindOnce(&DoubleFetcher::OnRequestComplete, base::Unretained(this),
+                     url),
+      kMaxDoubleFetchResponseSize);
+}
+
+void DoubleFetcher::OnRequestComplete(
+    GURL url,
+    std::optional<std::string> response_body) {
+  bool result = false;
+  auto* response_info = url_loader_->ResponseInfo();
+  if (response_info) {
+    auto response_code = response_info->headers->response_code();
+    if (!network::IsSuccessfulStatus(response_code)) {
+      if (response_code >= net::HttpStatusCode::HTTP_BAD_REQUEST &&
+          response_code < net::HttpStatusCode::HTTP_INTERNAL_SERVER_ERROR) {
+        // Only retry failures due to server error
+        // Mark as 'successful' if not a 5xx error, so we don't retry
+        result = true;
+      }
+      response_body = std::nullopt;
+    } else {
+      result = true;
+    }
+  }
+
+  auto request_data = request_queue_.NotifyRequestComplete(result);
+
+  if (request_data) {
+    const auto& request_dict = request_data->GetDict();
+    const auto* assoc_data = request_dict.Find(kAssociatedDataKey);
+    if (assoc_data) {
+      callback_.Run(url, *assoc_data, response_body);
```

Comment:
is it correct that we intentionally don't run the callback if anything failed (i.e. no `assoc_data`)?

Reply:
yes